streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -0,0 +1,178 @@
+            try {
+                final List<PartitionInfo> partitions = consumer.partitionsFor(topicPartition.topic());
+                if (partitions != null) {
+                    partitionInfo.put(topicPartition.topic(), partitions);
+                }
+            } catch (final TimeoutException e) {
+                throw new StreamsException(String.format("Could not fetch partition info for topic: %s before expiration of the configured request timeout",
+                                                         topicPartition.topic()));
+            }
+        }
+
+        if (!hasPartition(topicPartition)) {
+            throw new StreamsException(String.format("Store %s's change log (%s) does not contain partition %s",
+                                                     storeName, topicPartition.topic(), topicPartition.partition()));
+        }
+
+        log.debug("Took {} ms to validate that partition {} exists", time.milliseconds() - start, topicPartition);
+
+    }
+
+    @Override
+    public void register(final StateRestorer restorer) {
+        stateRestorers.put(restorer.partition(), restorer);
+    }
+
+    public void restore() {
+        final long start = time.milliseconds();
+        try {
+            if (!consumer.subscription().isEmpty()) {
+                throw new IllegalStateException(String.format("Restore consumer should have not subscribed to any partitions (%s) beforehand", consumer.subscription()));
+            }
+            final Map<TopicPartition, Long> endOffsets = consumer.endOffsets(stateRestorers.keySet());
+            consumer.assign(stateRestorers.keySet());
+
+            seek();
+
+            final Set<TopicPartition> partitions = new HashSet<>(stateRestorers.keySet());
+            while (!partitions.isEmpty()) {
+                final ConsumerRecords<byte[], byte[]> allRecords = consumer.poll(10);
+                final Iterator<TopicPartition> partitionIterator = partitions.iterator();
+                while (partitionIterator.hasNext()) {
+                    restorePartition(endOffsets, allRecords, partitionIterator);
+                }
+            }
+        } finally {
+            consumer.assign(Collections.<TopicPartition>emptyList());
+            log.debug("Took {} ms to restore active state", time.milliseconds() - start);
+        }
+    }
+
+    private void restorePartition(final Map<TopicPartition, Long> endOffsets,
+                                  final ConsumerRecords<byte[], byte[]> allRecords,
+                                  final Iterator<TopicPartition> partitionIterator) {
+        final TopicPartition topicPartition = partitionIterator.next();
+        final StateRestorer restorer = stateRestorers.get(topicPartition);
+        final Long endOffset = endOffsets.get(topicPartition);
+        final long pos = processNext(allRecords.records(topicPartition), restorer);
+        if (restorer.hasCompleted(pos, endOffset)) {
```

Comment:
If start pos < offset limit < end offset, then although no records beyond offset limit will be processed, they are still fetched until pos = end offset, and the pos will be recorded as restored offset. Is that right? Should we just stop fetching / processing on offset limit, and set the restored offset at offset limit?